compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3619,6 +3622,111 @@ public void visit(BLangStringTemplateLiteral stringTemplateLiteral) {
         resultType = types.checkType(stringTemplateLiteral, symTable.stringType, expType);
     }
 
+    @Override
+    public void visit(BLangRawTemplateLiteral rawTemplateLiteral) {
+        // First, ensure that the contextually expected type is compatible with the RawTemplate type.
+        // The RawTemplate type should have just two fields: strings and insertions. There shouldn't be any methods.
+        BType type = determineRawTemplateLiteralType(rawTemplateLiteral, expType);
+
+        if (type == symTable.semanticError) {
+            resultType = type;
+            return;
+        }
+
+        // Once we ensure the types are compatible, need to ensure that the types of the strings and insertions are
+        // compatible with the types of the strings and insertions fields.
+        BObjectType literalType = (BObjectType) type;
+        BType stringsType = literalType.fields.get("strings").type;
+
+        if (evaluateRawTemplateExprs(rawTemplateLiteral.strings, stringsType, INVALID_NUM_STRINGS,
+                                     rawTemplateLiteral.pos)) {
+            type = symTable.semanticError;
+        }
+
+        BType insertionsType = literalType.fields.get("insertions").type;
+
+        if (evaluateRawTemplateExprs(rawTemplateLiteral.insertions, insertionsType, INVALID_NUM_INSERTIONS,
+                                     rawTemplateLiteral.pos)) {
+            type = symTable.semanticError;
+        }
+
+        resultType = type;
+    }
+
+    private BType determineRawTemplateLiteralType(BLangRawTemplateLiteral rawTemplateLiteral, BType expType) {
+        // Contextually expected type is NoType when `var` is used.
+        // Therefore consider the literal as of type RawTemplate
+        if (expType == symTable.noType) {
+            return symTable.rawTemplateType;
+        }
+
+        BType type = types.checkType(rawTemplateLiteral, expType, symTable.rawTemplateType);
```

Comment:
Hmm yes, that does look wrong. I guess the correct way to do this would be to do what you suggested in https://github.com/ballerina-platform/ballerina-lang/pull/23705#discussion_r439572041 and not switch the order. Will change.

Reply:
Had a look at this. Instead of type checking against an inferred type, how about giving a different error here, instead of the standard incompatible types error? The reason is, to infer a type, we'll have to first individually type check the strings and insertions. But since now we are not type checking those two fields against a type, we'll be losing more accurate error messages we can give like in the following scenario.
```ballerina
type Temp abstract object {
    public [string, string, string] strings;
    public [int, float] insertions;
};

function testTypeChecking() {
    Temp rt = `Foo${8}Bar${"asdf"}`;
}
``` 
The error:
```
error: .::test.bal:7:29: incompatible types: expected 'float', found 'string'
```
If we infer a type and type check what we'd get is the standard incompatible message like:
```
error: .::test.bal:7:16: incompatible types: expected 'Temp', found 'abstract object {....}'
```

Reply:
Yeah. I also think a different error would be more explanatory here. It's a bit different from our usual errors, but I guess that's OK given that this is kinda different syntax to create a value of a specific type of object?

Reply:
Yeah, was thinking the same. Will type check with a different error message

Reply:
Changed to the following:
```
invalid literal for type 'string': raw templates can only be assigned to abstract subtypes of 'ballerina/lang.object:1.0.0:RawTemplate'
```